### PR TITLE
fix: Fix processing files from .gitignore

### PR DESCRIPTION
**Pull request by last-partizan**
Process only files returned by `git ls-files`.

Refs #711

### CODE REVIEW

File: seagoat/repository.py [hunk 1]
```diff
@@ -98,7 +103,7 @@ def analyze_files(self):
                     if not is_file_type_supported(filename):
                         continue
 
-                    if not (self.path / filename).exists():
+                    if filename not in git_files:
```
Comment (last-partizan) [inline seagoat/repository.py:106]: @kantord maybe we also should skip files excluded by ignorePatterns right here?

Comment (kantord) [reply seagoat/repository.py:106]: yes, but then we should also remove this logic:

```
    def _is_file_ignored(self, path: str):
        for pattern in self.config["server"]["ignorePatterns"]:
            if Path(path).match(pattern):
                return True

        return False
```

**Comment by kantord** [comment]
there are also some failing test cases, and it would also be nice to add some test case that verifies this new behavior (i. e. simulates some file that would have been previously added and won't be added after this change)

**Review by kantord — CHANGES_REQUESTED** [review]
see comments

File: tests/conftest.py [hunk 2]
```diff
@@ -164,12 +164,16 @@ def add_file_change_commit(
...
-            output_file.write(contents)
+        file_path = os.path.join(self.working_dir, file_name)
+
+        if contents is None:
+            os.unlink(file_path)
+            self.index.remove(file_name)
+        else:
+            with open(file_path, "w", encoding=encoding) as output_file:
+                output_file.write(contents)
+            self.index.add([file_name])
```
Comment (last-partizan) [inline tests/conftest.py:175]: Writing this test turned out trickier than i expected, but look at this.

We need to be able to commit deleted files, so i added this if/else block. But, maybe just for clarity i should rewrite this as separate "add_file_delete_commit"?

File: tests/test_queue.py [hunk 3]
```diff
@@ -55,7 +55,7 @@ def test_handle_get_stats(
 def test_important_files_are_analyzed_first(create_task_queue, mocker, repo):
     enqueue = mocker.patch("seagoat.queue.task_queue.TaskQueue.enqueue")
     create_task_queue()
-    sleep(0.5)
+    sleep(2.5)
```
Comment (last-partizan) [inline tests/test_queue.py:58]: Some other tests were failing because of this timeout, at least on my laptop. So I increased it.

**Comment by last-partizan** [comment]
Codacity is failing because we're using `check_output` with a variable? Can we ignore this, or i should hardcode full command like in other cases?

**Comment by last-partizan** [comment]
One other thing I noticed while testing it, patterns in `.gitignore` and our `ignorePatterns` is actually using different syntax, and probably we should switch to something like https://github.com/cpburnz/python-pathspec (Recommended by Claude.ai, or any other if you have something in mind. This recommendation seems good enough for me)

This would allow us to remove those two `git ls-files` calls, and rely on internal filtration.

File: tests/test_queue.py (hunk 3 above)
Comment (kantord) [reply tests/test_queue.py:58]: sure! but does it need to be increase all the way to 2.5s? if so, I'm ok with this change of course, it's important for the tests to run properly on any reasonable computer

**Comment by kantord** [comment]
> Codacity is failing because we're using `check_output` with a variable? Can we ignore this, or i should hardcode full command like in other cases?

yes! we can ignore that one

**Comment by kantord** [comment]
> One other thing I noticed while testing it, patterns in `.gitignore` and our `ignorePatterns` is actually using different syntax, and probably we should switch to something like https://github.com/cpburnz/python-pathspec (Recommended by Claude.ai, or any other if you have something in mind. This recommendation seems good enough for me)
> 
> This would allow us to remove those two `git ls-files` calls, and rely on internal filtration.

I am not sure if it's really a problem that the syntax is different in these cases :thinking: It might be ok. I am wondering if it's worth it to add an additional dependency for this purpose if the one built-in into Python works reasonably well.

However we might consider using `.full_match()` instead of `.match()` :thinking:  https://docs.python.org/3/library/pathlib.html#pathlib.PurePath.full_match

**Comment by kantord** [comment]
> This would allow us to remove those two git ls-files calls, and rely on internal filtration.

hmmm, I wonder which is more performant. I am thinking that git's built in ignore logic must be very efficient, so it might be better to rely on it as much as possible. In fact I'm wondering if it's actually possible to use git itself to add an additional ignore file :thinking: 



Would something like this work?

```bash
git ls-files --exclude-standard --exlude-from="path-to-anothe-exclude-file"
```

This way there could be a .seagoat_ignore file in addition to the .gitignore file.

The we could achieve what you mean: have only one format for specifying ignore files, while not requiring to add an additional dependency and actually remove code from SeaGOAT :thinking:  Albeit this would be a breaking change, because the ignorePatterns would be removed, but I'm ok with that

**Comment by kantord** [comment]
BTW one question: what should happen with untracked files? I think it would be important to include them in SeaGOAT results, albeit once they are tracked or explicitly ignored they should no longer show up in results (and I guess we can have a mechanism for "purging" the vector index of unused data, but that is another feature entirely, not relevant for this PR)

File: tests/conftest.py (hunk 2 above)
Comment (kantord) [reply tests/conftest.py:175]: yeah, I think `add_file_delete_commit` makes a lot of sense!

File: tests/test_queue.py (hunk 3 above)
Comment (last-partizan) [reply tests/test_queue.py]: Managed to get it down 2.0, but not lower.

**Comment by last-partizan** [comment]
> BTW one question: what should happen with untracked files? I think it would be important to include them in SeaGOAT results, albeit once they are tracked or explicitly ignored they should no longer show up in results (and I guess we can have a mechanism for "purging" the vector index of unused data, but that is another feature entirely, not relevant for this PR)

That's a great question. They definitely need to be tracked.

My initial thought was maybe we should use `rg` instead, we already using it and `rg --files` gives us list of files, including untracked. Then, i tested it in the shell agains `git ls-files`, and rg was slightly slower, but in python it's faster:

```python
In [3]: %time subprocess.check_output(['git', 'ls-files']); subprocess.check_output(['git', 'ls-files', '--deleted'])
CPU times: user 150 µs, sys: 10 ms, total: 10.2 ms
Wall time: 43.2 ms
Out[3]: b''

In [4]: %time subprocess.check_output(['rg', '--files']);
CPU times: user 2.59 ms, sys: 3.48 ms, total: 6.07 ms
Wall time: 31.4 ms
```

I'll try replacing `ls-files` with `rg` and see how it goes. And it support gitignore syntax for it's ignore rules, so we can remove our custom filtering altogether.

**Comment by last-partizan** [comment]
`rg` worked very well, but i'm leaving refactoring `ignoreFiles` for later. It's fast enough already.

I think, using `.seagot_ignore` instead of `ignorePatterns` would be great. (or `.sgignore`)

Too bad `--glob` is supposed only for positive matching. We could use something smart, like inverting each of our `ignorePatterns` and passing it to `--glob`, but i doubt this is a good idea.

But we can definitely pass a list of supported files to `--glob` and let `rg` do the rest.

**Pull request merged**